File: frontend/types/news.ts
/**
 * NewsItem (Production-Ready, Flexible + Safe)
 * -------------------------------------------
 * - Handles real-world backend inconsistencies
 * - Strong typing without breaking UI
 * - Compatible with filtering + charts
 */

export type EntityMap = Record<string, string[]>;

export type NewsItem = {
  id: number;
  title: string;
  summary: string;

  // Keywords for category + search
  keywords: string[];

  /**
   * Entities can come in 2 formats:
   * 1. Grouped (expected):
   *    { PERSON: ["Modi"], ORG: ["ICRA"] }
   *
   * 2. Flat (fallback from backend bugs / legacy):
   *    ["Modi", "India"]
   */
  entities?: EntityMap | string[];

  // Importance score (0–100)
  importance?: number;

  // Timestamps (optional depending on API)
  created_at?: string;
  published_at?: string;
};

File: backend/app/schema/news.py
from pydantic import BaseModel
from typing import List, Dict, Optional
from datetime import datetime


class NewsItemResponse(BaseModel):
    id: int
    title: str
    summary: str
    keywords: List[str]
    entities: Dict[str, List[str]]
    importance: float
    published_at: Optional[datetime]

    class Config:
        from_attributes = True


class NewsListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[NewsItemResponse]